### ISSUE TITLE
no-constant-condition reported error range is too broad

### ISSUE THREAD

**Issue by lydell** [eslint/eslint#9398]
**Tell us about your environment**

* **ESLint Version:** 4.8.0
* **Node Version:** 8.3.0
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "rules": {
    "no-constant-condition": "error"
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
if (true) {
  something();
}

something(true ? 1 : 0);
```

**What did you expect to happen?**

The reported error ranges should only include the constant conditions themselves. In the above cases, the words "true".

**What actually happened? Please include the actual, raw output from ESLint.**

The reported error ranges include the entire statement/expression that the constant condition is part of.

```json
$ eslint test.js --format json | jq
[
  {
    "filePath": "/home/lydell/bugs/eslint/test.js",
    "messages": [
      {
        "ruleId": "no-constant-condition",
        "severity": 2,
        "message": "Unexpected constant condition.",
        "line": 1,
        "column": 1,
        "nodeType": "IfStatement",
        "source": "if (true) {",
        "endLine": 3,
        "endColumn": 2
      },
      {
        "ruleId": "no-constant-condition",
        "severity": 2,
        "message": "Unexpected constant condition.",
        "line": 5,
        "column": 11,
        "nodeType": "ConditionalExpression",
        "source": "something(true ? 1 : 0);",
        "endLine": 5,
        "endColumn": 23
      }
    ],
    "errorCount": 2,
    "warningCount": 0,
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "source": "if (true) {\n  something();\n}\n\nsomething(true ? 1 : 0);\n"
  }
]
```

Expected:

```diff
@@ -7,11 +7,11 @@
         "severity": 2,
         "message": "Unexpected constant condition.",
         "line": 1,
-        "column": 1,
+        "column": 5,
         "nodeType": "IfStatement",
         "source": "if (true) {",
-        "endLine": 3,
-        "endColumn": 2
+        "endLine": 1,
+        "endColumn": 9
       },
       {
         "ruleId": "no-constant-condition",
@@ -22,7 +22,7 @@
         "nodeType": "ConditionalExpression",
         "source": "something(true ? 1 : 0);",
         "endLine": 5,
-        "endColumn": 23
+        "endColumn": 15
       }
     ],
     "errorCount": 2,
```

**Note:** I don't know how the reported `"source"` properties are supposed to work, so I can't tell whether they are correct or not.

Visual representation of the problem:

![screenshot from 2017-10-07 14-53-15](https://user-images.githubusercontent.com/2142817/31308016-4e871674-ab6f-11e7-828b-e0f40ed697d5.png)

Notice how much more than the constant conditions are underlined.

See also: https://github.com/steelbrain/linter-ui-default/issues/398#issuecomment-331632791 and onwards.

**Comment by not-an-aardvark** [comment]
Thanks for the report. I agree that it would be better to only report the condition here.

**Comment by i-ron-y** [comment]
It sounds like `line`, `column`, `endLine`, and `endColumn` show the location of the node in which the error is found. In the first example, the `nodeType` is `IfStatement`, and the location listed is the entire if-statement where the first `true` is found. In the second example, the `nodeType` is `ConditionalExpression`, and the location listed is the entire conditional expression where the second `true` is found.

**Comment by not-an-aardvark** [comment]
That's correct. I think this issue is occurring because the rule reports the entire `IfStatement` node when it should really just report the condition in the `if` statement. (The `line`, `column`, `endLine`, and `endColumn` are automatically added to the report based on the node that was reported.)

**Comment by VictorHom** [comment]
would the ideal case be: just reporting the position of the constant condition (so the `true`)?

so for the IfStatement, can we report with the `loc: {start: {line:nodestartPosition, column:node.test.range[0]}, end: {line:nodestartPosition, column:node.test.range[1]}}`?

haven't checked if there is a case that catches the correct position across the board, but I imagine they might be similar.

**Comment by not-an-aardvark** [comment]
We can just report `node.test`, and then all of that information will be automatically added to the report.
